clash-protocols-memmap/src/Protocols/MemoryMap.hs

```diff
@@ -0,0 +1,188 @@
+
+annotationSnd' :: (NFDataX (Fwd a), NFDataX (Fwd b), NFDataX (Bwd c)) => Circuit (a, BackwardAnnotated ann b) c -> ann
+annotationSnd' (Circuit f) = ann'
+  where
+    ((_, (ann', _)), _) = f (deepErrorX "")
+
+
+annotation ::
+  (KnownDomain dom, NFDataX (Fwd a), NFDataX (Bwd b)) =>
+  ((HiddenClockResetEnable dom) => Circuit (BackwardAnnotated ann a) b)
+  -> ann
+annotation circ = ann'
+  where
+    Circuit f = withClockResetEnable clockGen resetGen enableGen circ
+    ((ann', _), _) = f (deepErrorX "")
+
+annotationSnd ::
+  (KnownDomain dom, NFDataX (Fwd a), NFDataX (Fwd b), NFDataX (Bwd c)) =>
+  ((HiddenClockResetEnable dom) => Circuit (a, BackwardAnnotated ann b) c)
+  -> ann
+annotationSnd circ = ann'
+  where
+    Circuit f = withClockResetEnable clockGen resetGen enableGen circ
+    ((_, (ann', _)), _) = f (deepErrorX "")
+
+data Name = Name
+  { name :: String
+    -- ^ Name of the 'thing'. Used as an identifier in generated data structures.
+    --
+    -- TODO: Only allow very basic names here to make sure it can be used in all
+    --       common target languages. Provide a Template Haskell helper to make
+    --       sure the name is valid at compile time?
+  , description :: String
+  -- ^ Description of the 'thing'. Used in generated documentation.
+  }
+  deriving (Show, Eq, Ord)
+
+-- | Wrapper for \"things\" that have a name and a description. These are used
+-- to generate documentation and data structures for target languages.
+type Named a = (Name, a)
+type NamedLoc a = (Name, SrcLoc, a)
+
+type AbsoluteAddress = Maybe Address
+
+type DeviceName = String
+
+data MemoryMap = MemoryMap
+  { deviceDefs :: Map.Map DeviceName DeviceDefinition
+  , tree :: MemoryMapTree
+  }
+  deriving (Show)
+
+data DeviceDefinition = DeviceDefinition
+  { deviceName :: Name
+  , registers :: [NamedLoc Register]
+  , defLocation :: SrcLoc
+  }
+  deriving (Show)
+
+-- | A tree structure that describes the memory map of a device. Its definitions
+-- are using non-translatable constructs on purpose: Clash is currently pretty
+-- bad at propagating contants properly, so designers should only /produce/
+-- memory maps, not rely on constant folding to be able to extract addresses
+-- from them to use in their designs.
+data MemoryMapTree
+  = Interconnect SrcLoc AbsoluteAddress [(Address, Size, MemoryMapTree)]
+  | DeviceInstance SrcLoc AbsoluteAddress String DeviceName
+  deriving (Show)
+
+data Access
+  = ReadOnly
+  -- ^ Managers should only read from this register
+  | WriteOnly
+  -- ^ Managers should only write to this register
+  | ReadWrite
+  -- ^ Managers can read from and write to this register
+  deriving (Show)
+
+data Register = Register
+  { access :: Access
+  , address :: Address
+    -- ^ Address / offset of the register
+  , fieldType :: FieldType
+    -- ^ Type of the register. This is used to generate data structures for
+    -- target languages.
+  , fieldSize :: Size
+    -- ^ Size of the register in bytes
```

Comment:
Could we not derive this from the `FieldType`?

Reply:
In theory, yes, but it would involve a bunch of duplicated work. This field is meant to be filled by the `BitPackC`, I initially added a `sizeOf :: FieldType -> Size` function but it was pretty complex as it has to encode the same rules as `BitPackC` does. If we're fine with this duplication then I could add a function like that.